// react-router-async-component/src/index.js
import React, { Component } from 'react'

class Acomponent extends Component {
  constructor(props) {
    super(props)
    this.state = {
      view: null
    }
  }
  viewLoad() {
    const name = this.props.name || 'default'
    this.props.from.then(m => {
      this.setState({
        view: m[name]
      })
    }).catch(reason => {
      if(/Cannot find module/.test(reason)){
        this.props.notFound &&
        this.props.notFound.then && 
        this.props.notFound.then(m => {
          this.setState({
            view: m.default
          })
        })
      }
      else{
        this.props.onError &&
        this.props.onError.then && 
        this.props.onError.then(m => {
          this.setState({
            view: m.default
          })
        })
        console.error(reason)
      }
    })
  }
  componentDidMount() {
    this.viewLoad()
  }
  componentWillReceiveProps(nextProps) {
    if (this.props.from && this.props.from !== nextProps.from) {
      this.setState({
        view: null
      }, () => {
        this.viewLoad()
      })
    }
  }
  render() {
    if (this.state.view) {
      return React.createElement(this.state.view)
    }
    return null
  }
}

export default Acomponent
